src/tools/openapi.ts

```diff
@@ -0,0 +1,164 @@
+                "Do not replace variables in path, instead of, put them to the parameters object.",
+            },
+            method: { const: method, description: methodSpec.summary || methodSpec.description },
+            ...(methodSpec.requestBody?.content?.["application/json"]?.schema
+              ? {
+                  body: methodSpec.requestBody?.content?.["application/json"]?.schema,
+                }
+              : {}),
+            ...(methodSpec.parameters
+              ? {
+                  parameters: {
+                    type: "object",
+                    additionalProperties: false,
+                    required: methodSpec.parameters
+                      .filter((p: any) => p.required === true)
+                      .map((p: any) => p.name),
+                    properties: methodSpec.parameters.reduce(
+                      (acc: any, p: any) => ({
+                        ...acc,
+                        [p.name]: { ...p.schema, description: p.name },
+                      }),
+                      {},
+                    ),
+                  },
+                }
+              : {}),
+          },
+        })),
+      ),
+    } as const satisfies SchemaObject;
+  }
+
+  public readonly emitter: ToolEmitter<Record<string, any>, OpenAPIToolOutput, OpenAPIEvents> =
+    Emitter.root.child({
+      namespace: ["tool", "web", "openAPITool"],
+      creator: this,
+    });
+
+  static {
+    this.register();
+  }
+
+  public constructor(options: OpenAPIToolOptions) {
+    super(options);
+    this.openApiSchema = parse(options.openApiSchema);
+    if (!this.openApiSchema?.paths) {
+      throw new ValueError("Server is not specified!");
+    }
+  }
+
+  protected async _run(
+    input: any,
+    _options: Partial<BaseToolRunOptions>,
+    run: GetRunContext<typeof this>,
+  ) {
+    let path: string = input.path || "";
+    const url = new URL(this.openApiSchema.servers[0].url);
+    Object.keys(input.parameters ?? {}).forEach((key) => {
+      const value = input.parameters[key];
+      const newPath = path.replace(`{${key}}`, value);
+      if (newPath == path) {
+        url.searchParams.append(key, value);
+      } else {
+        path = newPath;
+      }
+    });
+    url.pathname = join(url.pathname, path);
+    // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
+    const headers: { [key: string]: string } = { Accept: "application/json" };
+    if (this.apiKey) {
+      headers["Authorization"] = `Bearer ${this.apiKey}`;
+    }
+    try {
+      const response = await fetch(url.toString(), {
+        body: !isEmpty(input.body) ? input.body : undefined,
+        method: input.method.toLowerCase(),
+        headers: headers,
+        signal: run.signal,
+      });
+      const text = await response.text();
+      return new OpenAPIToolOutput(response.status, response.statusText, text);
+    } catch {
+      throw new ToolError(`Request to ${url} failed.`);
```

Comment:
If you throw this error without any context, the agent cannot retry. Do this instead.

```ts
} catch (err) {
   throw new ToolError(`Request to ${url} has failed.`, [err]);
}
```